### PR TITLE
User Index Show Update Destroy

### PR DESCRIPTION
**Pull request by Andrii-Kozak**
**CHECK LIST**
- [x] specs added
- [x] specs passed
- [x] code coverage >90%
- [ ] DB indexes added if needed
- [x] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)

- [x] PR doesn't have conflicts with main branch
- [x] I've checked new feature as logged in and logged out user if needed
- [ ] "let!" is used in specs
- [x] screenshot of feature is added if task has UI

Знімок екрану з 2020-10-12 22-04-44

Screenshot (1)

### CODE REVIEW

File: app/javascript/stylesheets/application.scss [hunk 1]
```diff
@@ -1,6 +1,7 @@
-  @import "bootstrap"; 
+  @import "bootstrap";
   @import '@fortawesome/fontawesome-free';
   @import 'osbbs';
+  @import 'users';
```
Comment (mmarusyk) [inline app/javascript/stylesheets/application.scss:4]: Delete space before @import

File: db/seeds.rb [hunk 2]
```diff
@@ -4,4 +4,4 @@
 
 FactoryBot.create_list(:osbb, 100)
 
-FactoryBot.create_list(:company, 50)
+# FactoryBot.create_list(:company, 50)
```
Comment (mmarusyk) [inline db/seeds.rb:7]: Why did you comment this line?

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
+class UsersController < ApplicationController
+  before_action :user, only: %i[show edit update destroy]
+
+  def index
+    @users = User.all
```
Comment (pereslop) [inline app/controllers/users_controller.rb:5]: what will be if you have 1million users, you will load all them? will be usefull to add pagination here, and show about 40 users per page. Please read about will_paginate gem @chu4  is it OK?

Comment (pereslop) [reply app/controllers/users_controller.rb:5]: or kaminary, ect

File: app/javascript/stylesheets/application.scss (hunk 1 above)
Comment (Andrii-Kozak) [reply app/javascript/stylesheets/application.scss]: Thanks!
already fixed

File: db/seeds.rb (hunk 2 above)
Comment (Andrii-Kozak) [reply db/seeds.rb]: I forgot uncomment, fixed)

**Review by mmarusyk — APPROVED** [review]

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    redirect_to [:users]
+    flash[:danger] = "User profile has been deleted"
+  end
+
+  def update
+    if @user.update(user_params)
+      flash[:success] = "User profile \"#{@user.full_name}\"  updated"
+      redirect_to @user
+    else
+      render 'edit', warning: 'Invalid parameters for editing'
```
Comment (mmarusyk) [inline app/controllers/users_controller.rb:23]:
```suggestion
      flash.now[:warning] = 'Invalid parameters for editing!'
      render :edit
```

File: spec/controllers/users_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+RSpec.describe UsersController, type: :controller do
+  let!(:user) { create(:user) }
+  let!(:valid_params) { attributes_for :user }
+  let!(:invalid_params) { { first_name: ' ', last_name: ' ' } }
+
+  describe 'GET#index' do
+    it 'assigns users and renders template' do
+      get :index
+      expect(response).to have_http_status(:success)
```
Comment (IvanDanyliv) [inline spec/controllers/users_controller_spec.rb:11]: expect(assigns(:users)).to eq([user])

File: app/views/users/show.html.slim [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+      .user-item
+        span.nameing-user Phone:
+        span.insert-user = @user.mobile
+      .user-item
+        span.nameing-user Sex:
+        span.insert-user = @user.sex
+      nav
+        ul.nav.navbar-user
+         li.nav-item-user = link_to 'Edit User', edit_user_path, class: "btn btn-outline-warning"
+         li.nav-item-user = link_to 'Delete User',  @user, method: :delete, data: { confirm: 'Are you sure?' }, class: "btn btn-outline-danger"
```
Comment (mmarusyk) [inline app/views/users/show.html.slim:29]: Якщо аргументів більше два, то треба у власному рядку ставити. але то дрібниці:
```suggestion
li.nav-item-user
  = link_to 'Delete User', 
    @user,
    method: :delete,
    data: { confirm: 'Are you sure?' },
    class: "btn btn-outline-danger"
```

**Review by matsevich — APPROVED** [review]

File: app/models/user.rb [hunk 7]
```diff
@@ -8,25 +8,21 @@ class User < ApplicationRecord
...
   validates :last_name, presence: true, length: { maximum: 50 }
-  validates :email, format: { with: URI::MailTo::EMAIL_REGEXP }, uniqueness: true, length: { maximum: 255 }
-  validates :mobile, presence: true,
-                     numericality: true,
-                     length: { minimum: 10, maximum: 14 }
-  validates :sex, presence: true, length: { maximum: 10 }
-  validates :password, presence: true, length: { minimum: 5, maximum: 25 }
+  validates :email, format: { with: URI::MailTo::EMAIL_REGEXP },
+                    uniqueness: true, length: { maximum: 255 }
+  validates :mobile, numericality: true, allow_nil: true, length: { minimum: 10, maximum: 14 }
```
Comment (chu4) [inline app/models/user.rb:22]: why did you remove sex validation?

File: app/views/users/_form.html.slim [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+   = f.input :mobile, label: 'Phone number',
+                          placeholder: 'Phone number',
+                          input_html: { class: 'form-control' }
+
+  .form-group.user-form
+    = f.input :email, placeholder: 'user@domain.com',
+                               input_html: { class: 'form-control' }
+
+  .form-group.user-form
+    = f.input :sex, collection: User::SEX_TYPES,
```
Comment (chu4) [inline app/views/users/_form.html.slim:29]: maybe some default value here?

File: spec/models/user_spec.rb [hunk 9]
```diff
@@ -27,9 +27,9 @@
     end
 
     it "downcases email before saving" do
-      user.email = "example@gmail.com"
+      user.email = "exemple@gmail.com"
       user.save
-      expect(user.email).to eq("example@gmail.com")
+      expect(user.email).to eq("exemple@gmail.com")
```
Comment (chu4) [inline spec/models/user_spec.rb:32]: what does exemple mean?

File: app/models/user.rb (hunk 7 above)
Comment (Andrii-Kozak) [reply app/models/user.rb:22]: because user cannot register if sex is not set, and I also changed type of field to enum, so it does not need validation now.

File: app/views/users/_form.html.slim (hunk 8 above)
Comment (Andrii-Kozak) [reply app/views/users/_form.html.slim:29]: fixed

File: spec/models/user_spec.rb (hunk 9 above)
Comment (Andrii-Kozak) [reply spec/models/user_spec.rb:32]: fixed)

**Review by marikrebega — APPROVED** [review]

File: app/controllers/users_controller.rb (hunk 3 above)
Comment (Andrii-Kozak) [reply app/controllers/users_controller.rb]: done)

File: app/helpers/controller_helper.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+module ControllerHelper
```
Comment (vasya539) [inline app/helpers/controller_helper.rb:1]: thus you make code only more complicated

File: app/views/users/_form.html.slim [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+                           placeholder: 'First Name',
+                           input_html: { class: 'form-control' }
+
+  .form-group.user-form
+   = f.input :first_name, label: 'Last Name',
+                          placeholder: 'First Name',
+                          input_html: { class: 'form-control' }
+
+  .form-group.user-birth
+   = f.input :birthday, label: "Date of Birth :", as: :date, start_year: Date.today.year,
```
Comment (vasya539) [inline app/views/users/_form.html.slim:15]: alignment

File: db/migrate/20201006074043_add_fields_to_users.rb [hunk 12]
```diff
@@ -3,7 +3,7 @@ def change
     add_column :users, :mobile, :string
     add_column :users, :birthday, :date
     add_column :users, :avatar, :string
-    add_column :users, :sex, :string
+    add_column :users, :sex, :integer
```
[the remark this answers is not in the document]
Comment (Andrii-Kozak) [reply db/migrate/20201006074043_add_fields_to_users.rb]: fixed and created new migration)

File: app/views/users/_form.html.slim (hunk 11 above)
Comment (Andrii-Kozak) [reply app/views/users/_form.html.slim]: done)
